components/raftstore/src/store/fsm/peer.rs

```diff
@@ -737,159 +751,154 @@ where
-        for peer_id in to_be_removed.clone() {
-            let mut cc = eraftpb::ConfChangeV2::default();
-            let mut ccs = eraftpb::ConfChangeSingle::default();
-            ccs.set_change_type(eraftpb::ConfChangeType::RemoveNode);
-            ccs.set_node_id(peer_id);
-            cc.set_transition(eraftpb::ConfChangeTransition::Auto);
-            cc.mut_changes().push(ccs);
-            if let Err(e) = self.fsm.peer.raft_group.apply_conf_change(&cc) {
-                panic!("fail to apply conf change for unsafe recover {:?}", e);
+    }
+
+    fn finish_unsafe_recovery_destroy_peer(&mut self) {
+        if let Some(UnsafeRecoveryState::Destroy { counter, report_id }) =
+            self.fsm.peer.unsafe_recovery_state.take()
+        {
+            if counter.fetch_sub(1, Ordering::SeqCst) == 1 {
+                start_unsafe_recovery_report(&self.ctx.router, report_id);
             }
         }
-        self.fsm
-            .peer
-            .peer_heartbeats
-            .retain(|&k, _| new_peer_list.contains(&k));
-        self.fsm
-            .peer
-            .peers_start_pending_time
-            .retain(|&(k, _)| new_peer_list.contains(&k));
-        for peer in to_be_removed {
-            self.fsm.peer.remove_peer_from_cache(peer);
-        }
-        self.fsm.peer.post_split();
-        self.fsm.reset_hibernate_state(GroupState::Chaos);
-        self.register_raft_base_tick();
     }
 
-    fn on_unsafe_recovery_wait_apply(&mut self, counter: Arc<AtomicUsize>) {
+    fn on_unsafe_recovery_wait_apply(
+        &mut self,
+        shared_state: Arc<Mutex<UnsafeRecoveryWaitApplySharedState>>,
+    ) {
+        let target_index = if self.fsm.peer.force_leader.is_some() {
+            self.fsm.peer.raft_group.raft.raft_log.last_index()
+        } else {
+            self.fsm.peer.raft_group.raft.raft_log.committed
+        };
         info!(
             "unsafe recovery wait apply";
             "region_id" => self.region_id(),
             "peer_id" => self.fsm.peer_id(),
-            "target_index" => self.fsm.peer.raft_group.raft.raft_log.committed,
-            "task_counter" => counter.load(Ordering::Relaxed)
+            "force_leader" => self.fsm.peer.force_leader.is_some(),
+            "target_index" => target_index,
         );
 
         self.fsm.peer.unsafe_recovery_state = Some(UnsafeRecoveryState::WaitApply {
-            target_index: self.fsm.peer.raft_group.raft.raft_log.committed,
-            task_counter: counter,
+            target_index,
+            shared_state,
         });
         self.fsm
             .peer
             .unsafe_recovery_maybe_finish_wait_apply(self.ctx, /*force=*/ self.fsm.stopped);
     }
 
+    fn on_unsafe_recovery_fill_out_report(
+        &mut self,
+        shared_state: Arc<Mutex<UnsafeRecoveryFillOutReportSharedState>>,
+    ) {
+        let mut self_report = pdpb::PeerReport::default();
+        match self.ctx.engines.raft.get_raft_state(self.region_id()) {
```

Comment:
we can construct it from memory state, no need to get it from engine

Reply:
Done